basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/ListHeadTailNodeType.java

```diff
@@ -0,0 +1,93 @@
+package club.mondaylunch.gatos.basicnodes;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import club.mondaylunch.gatos.core.data.DataBox;
+import club.mondaylunch.gatos.core.data.DataType;
+import club.mondaylunch.gatos.core.data.ListDataType;
+import club.mondaylunch.gatos.core.graph.connector.NodeConnector;
+import club.mondaylunch.gatos.core.graph.type.NodeType;
+
+public class ListHeadTailNodeType extends NodeType.Process {
+    private static final DataType<ExtractionMode> EXTRACTION_MODE_TYPE = DataType.register("extraction_mode_type", ExtractionMode.class);
+    @Override
+    public Map<String, DataBox<?>> settings() {
+        return Map.of(
+            "extraction_mode", getHeadExtractionBox()
+        );
+    }
+
+    @Override
+    public Set<NodeConnector.Input<?>> inputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        return Set.of(
+            new NodeConnector.Input<>(nodeId, "input", ListDataType.GENERIC_LIST)
+        );
+    }
+
+    @Override
+    public Set<NodeConnector.Output<?>> outputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        var outType = inputTypes.isEmpty() ? DataType.ANY : inputTypes.get("input");
+        return Set.of(
+            new NodeConnector.Output<>(nodeId, "first", outType),
+            new NodeConnector.Output<>(nodeId, "rest", outType.listOf())
+        );
```

Comment:
Remember the type of 'input' is always a list. You're treating it as a scalar here. Also, why use a different way to get the output type here and in `compute`?

Also, a list of ANY is different to a generic list (this is a deficit of the type system, which I will fix in a later PR). For now, you need something like `outType == DataType.ANY ? ListDataType.GENERIC_LIST : outType.listOf()`

Reply:
is it not safe to assume there will only be 1 input to this node?

Reply:
rewritten